From dd3cb9289cc7e587f32127489b67b824ce0ebc53 Mon Sep 17 00:00:00 2001
From: zhouhuacai <zhouohuacai@oppo.com>
Date: Thu, 29 Apr 2021 17:23:05 +0800
Subject: [PATCH 10/15] ANDROID: vendor_hooks: tune reclaim inactive ratio

Add hooks to apply oem's memory reclaim optimization

Bug: 186669330
Bug: 470903229
Change-Id: I9a1c713bf3eea2823b2758ee44284a244b1ca7a1
Signed-off-by: zhouhuacai <zhouohuacai@oppo.com>
Signed-off-by: Tangquan Zheng <zhengtangquan@oppo.com>
Signed-off-by: zhaorou <v-zhaorou@oppo.com>
(cherry picked from commit 577f73412f579ef493e8e67795b3fd93f1a47dfd)
(cherry picked from commit 0e8c43272eaacf6a33117fdd721df6cb1877e5ec)
---
 drivers/android/vendor_hooks.c | 1 +
 include/trace/hooks/vmscan.h   | 3 +++
 mm/vmscan.c                    | 2 ++
 3 files changed, 6 insertions(+)

diff --git a/drivers/android/vendor_hooks.c b/drivers/android/vendor_hooks.c
index 93bbe26d2947..4d06fc8ce709 100644
--- a/drivers/android/vendor_hooks.c
+++ b/drivers/android/vendor_hooks.c
@@ -366,6 +366,7 @@ EXPORT_TRACEPOINT_SYMBOL_GPL(android_rvh_report_bug);
 EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_logbuf);
 EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_cpufreq_acct_update_power);
 EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_tune_swappiness);
+EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_tune_inactive_ratio);
 EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_logbuf_pr_cont);
 EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_calc_alloc_flags);
 EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_regmap_update);
diff --git a/include/trace/hooks/vmscan.h b/include/trace/hooks/vmscan.h
index ae87da63e814..ee3bc8036aac 100644
--- a/include/trace/hooks/vmscan.h
+++ b/include/trace/hooks/vmscan.h
@@ -25,6 +25,9 @@ DECLARE_HOOK(android_vh_check_folio_look_around_ref,
 DECLARE_HOOK(android_vh_tune_swappiness,
 	TP_PROTO(int *swappiness),
 	TP_ARGS(swappiness));
+DECLARE_HOOK(android_vh_tune_inactive_ratio,
+	TP_PROTO(unsigned long *inactive_ratio, int file),
+	TP_ARGS(inactive_ratio, file));
 DECLARE_HOOK(android_vh_shrink_folio_list,
 	TP_PROTO(struct folio *folio, bool dirty, bool writeback,
 		bool *activate, bool *keep),
diff --git a/mm/vmscan.c b/mm/vmscan.c
index 60d48861c73f..294acbe258aa 100644
--- a/mm/vmscan.c
+++ b/mm/vmscan.c
@@ -2397,6 +2397,8 @@ static bool inactive_is_low(struct lruvec *lruvec, enum lru_list inactive_lru)
 	else
 		inactive_ratio = 1;
 
+	trace_android_vh_tune_inactive_ratio(&inactive_ratio, is_file_lru(inactive_lru));
+
 	return inactive * inactive_ratio < active;
 }
 
-- 
2.50.1

